packages/eslint-plugin/src/rules/type-annotation-spacing.ts

```diff
@@ -32,6 +47,107 @@ const definition = {
   additionalProperties: false,
 };
 
+function createRules(options?: Config): WhitespaceRules {
+  const globals = {
+    ...(options?.before !== undefined ? { before: options.before } : {}),
+    ...(options?.after !== undefined ? { after: options.after } : {}),
+  };
+  const override = options?.overrides ?? {};
+  const colon = {
+    ...{ before: false, after: true },
+    ...globals,
+    ...override?.colon,
+  };
+  const arrow = {
+    ...{ before: true, after: true },
+    ...globals,
+    ...override?.arrow,
+  };
+
+  return {
+    colon: colon,
+    arrow: arrow,
+    variable: { ...colon, ...override?.variable },
+    property: { ...colon, ...override?.property },
+    parameter: { ...colon, ...override?.parameter },
+    returnType: { ...colon, ...override?.returnType },
+  };
+}
+
+function getNodeType(node?: TSESTree.Node): string {
+  return node?.type ?? '';
+}
+
+function typeIsArrowFunction(type: string): boolean {
+  return type === AST_NODE_TYPES.TSFunctionType;
+}
+
+function typeIsIdentifier(type: string): boolean {
+  return type === AST_NODE_TYPES.Identifier;
+}
+
+function typeIsProperty(type: string): boolean {
+  return [
+    AST_NODE_TYPES.ClassProperty,
+    AST_NODE_TYPES.FunctionExpression,
+    AST_NODE_TYPES.TSPropertySignature,
+    AST_NODE_TYPES.TSMethodSignature,
+  ].includes(type as AST_NODE_TYPES);
+}
+
+function typeIsReturnType(type: string): boolean {
+  return [
+    AST_NODE_TYPES.FunctionDeclaration,
+    AST_NODE_TYPES.ArrowFunctionExpression,
+  ].includes(type as AST_NODE_TYPES);
+}
+
+function typeIsVariable(type: string): boolean {
+  return type === AST_NODE_TYPES.VariableDeclarator;
+}
+
+function typeIsParameter(type: string): boolean {
+  return [
+    AST_NODE_TYPES.FunctionDeclaration,
+    AST_NODE_TYPES.TSFunctionType,
+    AST_NODE_TYPES.TSMethodSignature,
+    AST_NODE_TYPES.TSEmptyBodyFunctionExpression,
+  ].includes(type as AST_NODE_TYPES);
+}
+
+function getIdentifierRules(
+  rules: WhitespaceRules,
+  node?: TSESTree.Node,
```

Comment:
nit - avoid marking a parameter as optional if it's not actually optional.
An optional parameter can be omitted during the call, but a non-optional parameter with `undefined` in its union must be provided.
A very small delineation which helps make sure functions are called correctly.

```suggestion
  node: TSESTree.Node | undefined,
```

Reply:
Yes, that makes sense. Equating `undefined` with optional is a javascript habit, I guess...